peps/pep-0750.rst

```diff
@@ -0,0 +1,895 @@
+============================
+
+Python introduced f-strings in Python 3.6 with :pep:`498`. The grammar was
+then formalized in :pep:`701` which also lifted some restrictions. This PEP
+is based off of PEP 701.
+
+At nearly the same time PEP 498 arrived, :pep:`501` was written to provide
+"i-strings" -- that is, "interpolation template strings". The PEP was
+deferred pending further experience with f-strings. Work on this PEP was
+resumed by a different author in March 2023, introducing "t-strings" as template
+literal strings, and built atop PEP 701.
+
+The authors of this PEP consider tag strings as a generalization of the
+updated work in PEP 501.
+
+Motivation
+==========
+
+Python f-strings became very popular, very fast. The syntax was simple, convenient, and
+interpolated expressions had access to regular scoping rules. However, f-strings have
+two main limitations - expressions are eagerly evaluated, and interpolated values
+cannot be intercepted. The former means that f-strings cannot be re-used like templates,
+and the latter means that how values are interpolated cannot be customized.
+
+Templating in Python is currently achieved using packages like Jinja2 which bring their
+own templating languages for generating dynamic content. In addition to being one more
+thing to learn, these languages are not nearly as expressive as Python itself. This
+means that business logic, which cannot be expressed in the templating language, must be
+written in Python instead, spreading the logic across different languages and files.
+
+Likewise, the inability to intercept interpolated values means that they cannot be
+sanitized or otherwise transformed before being integrated into the final string. Here,
+the convenience of f-strings could be considered a liability. For example, a user
+executing a query with `sqlite3 <https://docs.python.org/3/library/sqlite3.html>`__
+may be tempted to use an f-string to embed values into their SQL expression instead of
+using the ``?`` placeholder and passing the values as a tuple to avoid an
+`SQL injection attack <https://en.wikipedia.org/wiki/SQL_injection>`__.
+
+Tag strings address both these problems by extending the f-string syntax to provide
+developers access to the string and its interpolated values before they are combined. In
+doing so, tag strings may be interpreted in many different ways, opening up the
+possibility for DSLs and other custom string processing.
+
+Proposal
+========
+
+This PEP proposes customizable prefixes for f-strings. These f-strings then
+become a "tag string": an f-string with a "tag function." The tag function is
+a callable which is given a sequence of arguments for the parsed tokens in
+the string.
+
+Here's a very simple example. Imagine we want a certain kind of string with
+some custom business policies: uppercase the value and add an exclamation point.
+
+Let's start with a tag string which simply returns a static greeting:
+
+.. code-block:: python
+
+    def greet():
```

Comment:
Should this one also take `*args` even if unused ?

Reply:
It's not only for consistency. Not adding it is a `TypeError`.

```python
>>> def greet():
...     return "Hello!"
...
>>> greet"Hi!"
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    greet"Hi!"
TypeError: greet() takes 0 positional arguments but 1 was given
```

Reply:
Maybe add that explicitely arround line 260, after

```
When the tag string is evaluated, the tag must have a binding, or a ``NameError``
is raised; and it must be a callable, or a ``TypeError`` is raised. This behavior
follows from the de-sugaring of:
```

Reply:
Changed to `greet(*args)`, with accompanying discussion by @pauleveritt